{
  "cells": [
    {
      "cell_type": "markdown",
      "id": "JZNTi8ZL8XqC",
      "metadata": {
        "id": "JZNTi8ZL8XqC"
      },
      "source": [
        "# ColabBE: automated base editing analysis for ClinVar variants\n",
        "\n",
        "Written by Angus Li with contributions from Alvin Hsu\n",
        "\n",
        "Editing predictions based on BE-Hive by Max W. Shen (https://www.crisprbehive.design/), an implementation of machine learning models for base editing outcomes described in:\n",
        "\n",
        "**Arbab M**, **Shen MW**, Mok B, et al. Determinants of Base Editing Outcomes from Target Library Analysis and Machine Learning. *Cell*. 2020;182(2):463-480.e30. doi:10.1016/j.cell.2020.05.037\n",
        "\n",
        "Cas9 variant recommendations based on HT-PAMDA data courtesy of Rachel E. Silverstein and Benjamin P. Kleinstiver, as described in:\n",
        "\n",
        "**Silverstein RA**, Kim N, Kroell AS, et al. Custom CRISPR-Cas9 PAM variants via scalable engineering and machine learning. *Nature*. Published online April 22, 2025. doi:10.1038/s41586-025-09021-y\n",
        "\n",
        "For coding mutations, ColabBE analyzes base editing strategies that are expected to correct the mutated codon to any codon for the reference amino acid. For non-coding mutations, ColabBE analyzes base editing strategies that are expected to revert the mutated base to the reference sequence.\n",
        "\n",
        "**Instructions**:\n",
        "\n",
        "- First run the section titled \"Install/import libraries and define functions\" by clicking the \"Play\" button to the left.\n",
        "- The optional \"Search ClinVar for ClinVar IDs\" section can be used to find ClinVar IDs of interest using more general search terms.\n",
        "- Specify all parameters under \"Perform analysis\" then start analysis by clicking the \"Play\" button to the left.\n",
        "\n",
        "**Results interpretation**:\n",
        "\n",
        "- When analysis is finished, a zip archive is created, containing all output files.\n",
        "- For each ClinVar ID with potential corrective strategies, two files are provided: `...predicted base editing statistics.xlsx` and `...predicted sequence distributions.xlsx`.\n",
        "-  `...predicted base editing statistics.xlsx` contains the editor, 20-nucleotide spacer, and 4-nucleotide PAM for each analyzed base editing strategy, as well as the fraction of predicted perfect correction, Z-score, percentile, and predicted target efficiency given the specified `average_editing_efficiency`. For coding mutations, perfect correction is defined as the fraction of **edited** sequencing reads for which the corresponding translated protein sequence is reverted to the same as the reference sequence. For non-coding mutations, perfect correction is defined as the fraction of **edited** sequencing reads for which the DNA sequence exactly matches the reference sequence. Cas9 variant recommendations and corresponding HT-PAMDA scores are also provided based on the PAM sequence.\n",
        "- `...predicted sequence distributions.xlsx` contains the predicted fraction of sequences resulting from each analyzed editing strategy. Each editor, spacer, and PAM combination is a separate sheet in the file.\n",
        "- A file named `Summary of editor candidates above threshold.xlsx` is also provided. This is a listing of all editor candidates, with their corresponding ClinVar IDs and variant names, with predicted perfect correction above the cutoff specified in `threshold`.\n",
        "- A file named `Variants with editor candidates above threshold.xlsx` is provided. This lists all analyzed variants with at least one editor candidate with predicted perfect correction above the cutoff specified in `threshold`.\n",
        "- All output files are also stored in a timestamped folder on your Google Drive. This can be deleted after analysis is finished if space is limited."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "69eed071",
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "69eed071",
        "outputId": "b0a706d2-921e-4f2a-88e3-158a5ea1bae6"
      },
      "outputs": [],
      "source": [
        "# Install/import libraries and define functions\n",
        "\n",
        "import requests\n",
        "from Bio import Seq\n",
        "from Bio import Entrez\n",
        "import io\n",
        "import pandas as pd\n",
        "from datetime import datetime\n",
        "import os\n",
        "import subprocess\n",
        "import re\n",
        "pam_dict = pd.read_excel('PAM_scores.xlsx', index_col = 'PAM').to_dict(orient='index')\n",
        "\n",
        "def cvIDtoHg38Coords(cvID):\n",
        "    ASM_MAP = {\"NCBI36\": \"hg18\", \"GRCh37\": \"hg19\", \"GRCh38\": \"hg38\", \"T2T-CHM13v2.0\": \"hs1\"}\n",
        "    resp = requests.get(\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?\",\n",
        "                      params = {\"db\": \"clinvar\", \"id\": cvID, \"retmode\": \"json\"})\n",
        "    if not resp.ok: raise Exception(\"Failed to query ClinVar\")\n",
        "    data = resp.json()\n",
        "    if \"result\" not in data or cvID not in data['result']: raise Exception(f'Invalid ClinVar ID: {cvID}')\n",
        "    varData = data[\"result\"][cvID][\"variation_set\"][0]\n",
        "    spdi = varData[\"canonical_spdi\"]\n",
        "    [start, ref, mut] = spdi.split(\":\")[1:]\n",
        "    coord = [entry for entry in varData[\"variation_loc\"] if entry['status'] == 'current'][0]\n",
        "    [chrom, assembly_name] = [coord['chr'], coord['assembly_name']]\n",
        "    return {'coords': {'assembly': ASM_MAP[assembly_name],\n",
        "                                  'chrom': \"chr\" + chrom,\n",
        "                                  'pos': start },\n",
        "                        'alleles': { 'vName': varData[\"variation_name\"],\n",
        "                                   'ref': ref,\n",
        "                                   'mut': mut },\n",
        "                        'gene': data[\"result\"][cvID][\"gene_sort\"]}\n",
        "\n",
        "def coordsToRefSeq(coords):\n",
        "    if 'start' in coords and 'end' in coords:\n",
        "        start = coords['start']\n",
        "        end = coords['end']\n",
        "    else:\n",
        "        start = int(coords['pos'])\n",
        "        end = start + 1\n",
        "    query = {'track': 'ccdsGene', 'genome': coords['assembly'], 'chrom': coords['chrom'],\n",
        "                 'start': start, 'end': end}\n",
        "    resp = requests.get(\"https://api.genome.ucsc.edu/getData/track\", params = query)\n",
        "    if not resp.ok: raise Exception(f'Failed to get RefSeq annotations from UCSC: {resp.status_code} {resp.text}')\n",
        "    data = resp.json()\n",
        "    if data['ccdsGene']: return data['ccdsGene'][0]\n",
        "    return None\n",
        "\n",
        "def fetchSequenceFromCoords(coords, contextLen):\n",
        "    [assembly, chrom, pos] = [coords['assembly'], coords['chrom'], int(coords['pos'])]\n",
        "    startPos = pos - contextLen\n",
        "    endPos = pos + contextLen\n",
        "    query = {'genome': assembly, 'chrom': chrom, 'start': startPos, 'end': endPos}\n",
        "    resp = requests.get('https://api.genome.ucsc.edu/getData/sequence', params = query)\n",
        "    if not resp.ok: raise Exception('Failed to fetch sequence from UCSC Genome Browser')\n",
        "    data = resp.json()\n",
        "    return data['dna']\n",
        "\n",
        "def getContextExonTranslations(geneData, target, contextLen):\n",
        "    target = int(target)\n",
        "    contextStart = target - contextLen\n",
        "    contextEnd = target + contextLen\n",
        "    exonStarts = [int(n) for n in geneData[\"exonStarts\"].split(',') if n.isdigit() and int(n) != 0]\n",
        "    exonEnds = [int(n) for n in geneData[\"exonEnds\"].split(',') if n.isdigit() and int(n) != 0]\n",
        "    exonFrames = [int(n) for n in geneData[\"exonFrames\"].split(',') if n.isdigit()]\n",
        "    if geneData['strand'] == '+':\n",
        "        exonStarts[0] = geneData['cdsStart']\n",
        "        exonEnds[-1] = geneData['cdsEnd']\n",
        "        exonFrames[0] = 0\n",
        "    else:\n",
        "        exonStarts[0] = geneData['cdsStart']\n",
        "        exonEnds[-1] = geneData['cdsEnd']\n",
        "        exonFrames[-1] = 0\n",
        "    contentExons = []\n",
        "    for i in range(len(exonStarts)):\n",
        "        if exonStarts[i] <= contextEnd and exonEnds[i] >= contextStart:\n",
        "            exonNumber = i + 1 if geneData['strand'] == '+' else len(exonStarts) - i\n",
        "            startOffset = max(0, exonStarts[i] - contextStart)\n",
        "            endOffset = min(exonEnds[i], contextEnd) - contextStart\n",
        "            adjustedFrame = exonFrames[i]\n",
        "            if exonStarts[i] < contextStart and geneData['strand'] == '+':\n",
        "                distanceFromContextStart = contextStart - exonStarts[i]\n",
        "                adjustedFrame = (exonFrames[i] + distanceFromContextStart) % 3\n",
        "            elif exonEnds[i] > contextEnd and geneData['strand'] == '-':\n",
        "                distanceFromContextStart = exonEnds[i] - contextEnd\n",
        "                adjustedFrame = (exonFrames[i] + distanceFromContextStart) % 3\n",
        "            contentExons.append({'name': f'{geneData[\"name2\"]} Exon {exonNumber}',\n",
        "                                 'start': startOffset, 'end': endOffset, 'direction': geneData['strand'],\n",
        "                                 'frame': (3 - adjustedFrame) % 3})\n",
        "    return contentExons\n",
        "\n",
        "def testReversion(mutCodon, wtAA, verbose = True):\n",
        "    candidates = []\n",
        "    abe_plus = mutCodon.replace('A', 'G')\n",
        "    if Seq.translate(abe_plus) == wtAA:\n",
        "        if verbose: print(f'{mutCodon} could be corrected to {abe_plus}')\n",
        "        candidates.append(['ABE8e', '+'])\n",
        "    cbe_plus = mutCodon.replace('C', 'T')\n",
        "    if Seq.translate(cbe_plus) == wtAA:\n",
        "        if verbose: print(f'{mutCodon} could be corrected to {cbe_plus}')\n",
        "        candidates.append(['BE4', '+'])\n",
        "    abe_minus = mutCodon.replace('T', 'C')\n",
        "    if Seq.translate(abe_minus) == wtAA:\n",
        "        if verbose: print(f'{mutCodon} could be corrected to {abe_minus}')\n",
        "        candidates.append(['ABE8e', '-'])\n",
        "    cbe_minus = mutCodon.replace('G', 'A')\n",
        "    if Seq.translate(cbe_minus) == wtAA:\n",
        "        if verbose: print(f'{mutCodon} could be corrected to {cbe_minus}')\n",
        "        candidates.append(['BE4', '-'])\n",
        "    return candidates\n",
        "\n",
        "def isTransition(allele, verbose = True):\n",
        "    if allele['mut'] == 'A' and allele['ref'] == 'G':\n",
        "        if verbose: print('A could be corrected to G')\n",
        "        return ['ABE8e', '+']\n",
        "    elif allele['mut'] == 'C' and allele['ref'] == 'T':\n",
        "        if verbose: print('C could be corrected to T')\n",
        "        return ['BE4', '+']\n",
        "    elif allele['mut'] == 'T' and allele['ref'] == 'C':\n",
        "        if verbose: print('T could be corrected to C')\n",
        "        return ['ABE8e', '-']\n",
        "    elif allele['mut'] == 'G' and allele['ref'] == 'A':\n",
        "        if verbose: print('G could be corrected to A')\n",
        "        return ['BE4', '-']\n",
        "    return False\n",
        "\n",
        "def behive(seq50,start,contextLen,editor,pred_efficiency,correctExon,strand=None):\n",
        "    frame = None\n",
        "    exonStart = 0\n",
        "    exonEnd = 2 * contextLen\n",
        "    if correctExon:\n",
        "        frame = correctExon['frame']\n",
        "        exonStart = correctExon['start']\n",
        "        exonEnd = correctExon['end']\n",
        "        if correctExon['direction'] == '+':\n",
        "            if strand == '+': frame_shifted = (exonStart + frame - start) % 3 + 1\n",
        "            else: frame_shifted = (-1*frame - ((exonEnd-exonStart)%3) + (2*contextLen-exonEnd)+start) % 3 + 1\n",
        "        elif correctExon['direction'] == '-':\n",
        "            if strand == '-': frame_shifted = (-1 * exonStart + ((exonEnd-exonStart)%3)+ frame + start) % 3 + 1\n",
        "            else: frame_shifted = (-1*frame  - (2*contextLen-exonEnd)-start) % 3 + 1\n",
        "    else:\n",
        "        frame_shifted = 1\n",
        "        strand = '+'\n",
        "\n",
        "    URL = 'https://www.crisprbehive.design'\n",
        "\n",
        "    dash_update = f'{URL}/_dash-update-component'\n",
        "\n",
        "    payload = {\n",
        "    \"output\":\"S_csv_download_link.href\",\n",
        "    \"changedPropIds\":[\n",
        "        \"S_hidden_pred_signal_bystander.children\"\n",
        "    ],\n",
        "    \"inputs\":[\n",
        "        {\n",
        "            \"id\":\"S_hidden_pred_signal_bystander\",\n",
        "            \"property\":\"children\",\n",
        "            \"value\":f\"{seq50},{editor},mES\"\n",
        "        },\n",
        "        {\n",
        "            \"id\":\"S_hidden_chosen_aa_frame\",\n",
        "            \"property\":\"children\",\n",
        "            \"value\":f\"{frame_shifted},{strand}\"\n",
        "        }\n",
        "    ]\n",
        "    }\n",
        "\n",
        "    headers = {\n",
        "        \"Content-Type\": \"application/json\"\n",
        "    }\n",
        "\n",
        "    resp = requests.post(dash_update, headers=headers, json=payload)\n",
        "    href = resp.json()['response']['props']['href']\n",
        "    resp = requests.get(f'{URL}{href}')\n",
        "    df = pd.read_csv(io.StringIO(resp.text), index_col=0)\n",
        "\n",
        "    payload = {\n",
        "        \"output\":\"S_efficiency_longtext.children\",\n",
        "        \"changedPropIds\":[\n",
        "            \"S_hidden_pred_signal_efficiency.children\"\n",
        "        ],\n",
        "        \"inputs\":[\n",
        "            {\n",
        "                \"id\":\"S_slider_efficiency_mean\",\n",
        "                \"property\":\"value\",\n",
        "                \"value\":pred_efficiency\n",
        "            },\n",
        "            {\n",
        "                \"id\":\"S_hidden_pred_signal_efficiency\",\n",
        "                \"property\":\"children\",\n",
        "                \"value\":f\"{seq50},{editor},mES\"\n",
        "            }\n",
        "        ]\n",
        "    }\n",
        "    resp = requests.post(dash_update, headers=headers, json=payload)\n",
        "    href = resp.json()\n",
        "    for block in href['response']['props']['children']:\n",
        "        s = block['props']['children']\n",
        "        if isinstance(s, str):\n",
        "            if 'Predicted Z-score:' in s:\n",
        "                z = float(s[19:])\n",
        "            elif 'Percentile:' in s:\n",
        "                percentile = float(s[-4:])\n",
        "            elif 'then this target\\'s efficiency is' in s:\n",
        "                target_efficiency = float(s[-6:-2])\n",
        "    return {'df':df, 'z':z, 'percentile':percentile, 'target_efficiency':target_efficiency}\n",
        "\n",
        "def analyze(cvId, verbose = True, pred_efficiency = 0.5):\n",
        "    contextLen = 35\n",
        "    results = {}\n",
        "    analysis = {}\n",
        "    stats_dict = []\n",
        "    a = cvIDtoHg38Coords(cvId)\n",
        "    vname = a['alleles']['vName']\n",
        "    print(f'ClinVar ID: {cvId}')\n",
        "    print(f'Variant name: {vname}')\n",
        "    if len(a['alleles']['ref']) != 1 or len(a['alleles']['mut']) != 1:\n",
        "        print(f'The mutation with ClinVar ID {cvId} is not a single nucleotide variant.')\n",
        "        return[analysis, stats_dict]\n",
        "    geneData = coordsToRefSeq(a['coords'])\n",
        "    seq = fetchSequenceFromCoords(a['coords'], contextLen)\n",
        "    seq = seq.upper()\n",
        "    mut = seq[:contextLen] + a['alleles']['mut'] + seq[contextLen+1:]\n",
        "    if geneData: c = getContextExonTranslations(geneData, a['coords']['pos'], contextLen)\n",
        "    else: c = {}\n",
        "    correctExon = None\n",
        "    for exon in c:\n",
        "        seqSlice = seq[exon['start']:exon['end']]\n",
        "        mutSlice = mut[exon['start']:exon['end']]\n",
        "        if exon['direction'] == '-':\n",
        "            seqSlice = Seq.reverse_complement(seqSlice)\n",
        "            mutSlice = Seq.reverse_complement(mutSlice)\n",
        "        seqTranslation = Seq.translate(seqSlice[exon['frame']:len(seqSlice)-((len(seqSlice)-exon['frame'])%3)])\n",
        "        mutTranslation = Seq.translate(mutSlice[exon['frame']:len(seqSlice)-((len(seqSlice)-exon['frame'])%3)])\n",
        "        if verbose:\n",
        "            print('WT:')\n",
        "            print(seqSlice)\n",
        "            print(' ' * exon['frame'] + ''.join([' ' + seqTranslation[aa] + ' ' for aa in range(len(seqTranslation))]))\n",
        "            print(Seq.complement(seqSlice))\n",
        "            print('Variant:')\n",
        "            print(mutSlice)\n",
        "            print(' ' * exon['frame'] + ''.join([' ' + mutTranslation[aa] + ' ' for aa in range(len(mutTranslation))]))\n",
        "            print(Seq.complement(mutSlice))\n",
        "        if contextLen < exon['end'] and contextLen > exon['start']:\n",
        "            correctExon = exon\n",
        "            if exon['direction'] == '+':\n",
        "                pos = contextLen - exon['start']\n",
        "            elif exon['direction'] == '-':\n",
        "                pos = (exon['end'] - 1) - contextLen\n",
        "            wtCodon = seqSlice[pos - ((pos - exon['frame']) % 3):pos - ((pos - exon['frame']) % 3) + 3]\n",
        "            mutCodon = mutSlice[pos - ((pos - exon['frame']) % 3):pos - ((pos - exon['frame']) % 3) + 3]\n",
        "            wtAA = Seq.translate(wtCodon)\n",
        "            mutAA = Seq.translate(mutCodon)\n",
        "            if verbose: print(f'The wild-type codon {wtCodon}, coding for {wtAA}, is mutated to {mutCodon}, coding for {mutAA}')\n",
        "    if correctExon:\n",
        "        candidates = testReversion(mutCodon, wtAA, verbose)\n",
        "        if candidates:\n",
        "            for candidate in candidates:\n",
        "                [editor, strand] = candidate\n",
        "                if strand == correctExon['direction']:\n",
        "                    for start in range(17):\n",
        "                        seq50 = mut[start:start+50]\n",
        "                        if verbose: print(f'Sending {seq50} (top strand) to BE-Hive with editor {editor}')\n",
        "                        results[(cvId, start, editor, strand)] = behive(seq50, start, contextLen, editor, pred_efficiency, correctExon, strand)\n",
        "                else:\n",
        "                    mut_rc = Seq.reverse_complement(mut)\n",
        "                    for start in range(17):\n",
        "                        seq50 = mut_rc[start:start+50]\n",
        "                        if verbose: print(f'Sending {seq50} (bottom strand) to BE-Hive with editor {editor}')\n",
        "                        results[(cvId, start, editor, strand)] = behive(seq50, start, contextLen, editor, pred_efficiency, correctExon, strand)\n",
        "            seqSlice = seq[correctExon['start']:correctExon['end']]\n",
        "            if correctExon['direction'] == '-':\n",
        "                seqSlice = Seq.reverse_complement(seqSlice)\n",
        "            seqTranslation = Seq.translate(seqSlice[correctExon['frame']:len(seqSlice)-((len(seqSlice)-correctExon['frame'])%3)])\n",
        "            for entry in results:\n",
        "                (cvId, start, editor, strand) = entry\n",
        "                df = results[entry]['df']\n",
        "                z = results[entry]['z']\n",
        "                percentile = results[entry]['percentile']\n",
        "                target_efficiency = results[entry]['target_efficiency']\n",
        "                perfect_correction = 0\n",
        "                if strand == correctExon['direction']:\n",
        "                    spacer = mut[start+20:start+40]\n",
        "                    pam = mut[start+40:start+44]\n",
        "                else:\n",
        "                    mut_rc = Seq.reverse_complement(mut)\n",
        "                    spacer = mut_rc[start+20:start+40]\n",
        "                    pam = mut_rc[start+40:start+44]\n",
        "                params = (editor, spacer, pam)\n",
        "                analysis[params] = {}\n",
        "                for p, genotype in zip(df['Predicted frequency'], df['Genotype']):\n",
        "                    if strand == correctExon['direction']:\n",
        "                        mut_edited = mut[:start] + genotype + mut[start+50:]\n",
        "                    else:\n",
        "                        mut_rc_edited = mut_rc[:start] + genotype + mut_rc[start+50:]\n",
        "                        mut_edited = Seq.reverse_complement(mut_rc_edited)\n",
        "                    mutSlice = mut_edited[correctExon['start']:correctExon['end']]\n",
        "                    if correctExon['direction'] == '-': mutSlice = Seq.reverse_complement(mutSlice)\n",
        "                    mutTranslation = Seq.translate(mutSlice[correctExon['frame']:len(seqSlice)-((len(seqSlice)-correctExon['frame'])%3)])\n",
        "                    if seqTranslation == mutTranslation:\n",
        "                        perfect_correction += p\n",
        "                    if mutTranslation not in analysis[params]: analysis[params][mutTranslation] = 0\n",
        "                    analysis[params][mutTranslation] += p\n",
        "                stats_dict.append({'ClinVar ID': cvId, 'Variant name': vname, 'Editor': editor, 'Spacer': spacer, 'PAM': pam, 'Perfect correction': perfect_correction,\n",
        "                                     'Z-score': z, 'Percentile': percentile, 'Target efficiency': target_efficiency})\n",
        "        else: print(f'The coding mutation with ClinVar ID {cvId} is unlikely to be correctable via canonical base editing. Try prime editing instead.')\n",
        "    else:\n",
        "        transition = isTransition(a['alleles'])\n",
        "        if transition:\n",
        "            [editor, strand] = transition\n",
        "            if strand == '+':\n",
        "                for start in range(17):\n",
        "                    seq50 = mut[start:start+50]\n",
        "                    if verbose: print(f'Sending {seq50} (top strand) to BE-Hive with editor {editor}')\n",
        "                    results[(cvId, start, editor, strand)] = behive(seq50, start, contextLen, editor, 0.5, correctExon)\n",
        "            elif strand == '-':\n",
        "                for start in range(17):\n",
        "                    seq50 = Seq.reverse_complement(mut)[start:start+50]\n",
        "                    if verbose: print(f'Sending {seq50} (bottom strand) to BE-Hive with editor {editor}')\n",
        "                    results[(cvId, start, editor, strand)] = behive(seq50, start, contextLen, editor, 0.5, correctExon)\n",
        "            for entry in results:\n",
        "                (cvId, start, editor, strand) = entry\n",
        "                df = results[entry]['df']\n",
        "                z = results[entry]['z']\n",
        "                percentile = results[entry]['percentile']\n",
        "                target_efficiency = results[entry]['target_efficiency']\n",
        "                perfect_correction = 0\n",
        "                if strand == '+':\n",
        "                    spacer = mut[start+20:start+40]\n",
        "                    pam = mut[start+40:start+44]\n",
        "                elif strand == '-':\n",
        "                    mut_rc = Seq.reverse_complement(mut)\n",
        "                    spacer = mut_rc[start+20:start+40]\n",
        "                    pam = mut_rc[start+40:start+44]\n",
        "                params = (editor, spacer, pam)\n",
        "                analysis[params] = {}\n",
        "                for p, genotype in zip(df['Predicted frequency'], df['Genotype']):\n",
        "                    if strand == '+':\n",
        "                        mut_edited = mut[:start] + genotype + mut[start+50:]\n",
        "                    elif strand == '-':\n",
        "                        mut_rc_edited = mut_rc[:start] + genotype + mut_rc[start+50:]\n",
        "                        mut_edited = Seq.reverse_complement(mut_rc_edited)\n",
        "                    if seq == mut_edited:\n",
        "                        perfect_correction += p\n",
        "                    if mut_edited not in analysis[params]: analysis[params][mut_edited] = 0\n",
        "                    analysis[params][mut_edited] += p\n",
        "                stats_dict.append({'ClinVar ID': cvId, 'Variant name': vname, 'Editor': editor, 'Spacer': spacer, 'PAM': pam, 'Perfect correction': perfect_correction,\n",
        "                                   'Z-score': z, 'Percentile': percentile, 'Target efficiency': target_efficiency})\n",
        "        else: print(f'The non-coding mutation with ClinVar ID {cvId} is not accesible to canonical base editing. Try prime editing instead.')\n",
        "    print('Done')\n",
        "    return [analysis, stats_dict, vname]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "_MorcrwER0z2",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 147
        },
        "id": "_MorcrwER0z2",
        "outputId": "475817ac-a63a-4c37-f2cf-96508bcb2d6c"
      },
      "outputs": [
        {
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "summary": "{\n  \"name\": \"df\",\n  \"rows\": 1,\n  \"fields\": [\n    {\n      \"column\": \"ClinVar ID\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"986214\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"title\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"NM_001244008.2(KIF1A):c.746T>C (p.Leu249Pro)\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"gene_sort\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"KIF1A\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"chr_sort\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"02\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"location_sort\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"00000000000240783791\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"obj_type\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"single nucleotide variant\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"protein_change\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"L249P\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"germline_classification\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"Pathogenic/Likely pathogenic\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}",
              "type": "dataframe",
              "variable_name": "df"
            },
            "application/vnd.google.colaboratory.module+javascript": "\n      import \"https://ssl.gstatic.com/colaboratory/data_table/e523c247d1e24a05/data_table.js\";\n\n      const table = window.createDataTable({\n        data: [[{\n            'v': 0,\n            'f': \"0\",\n        },\n\"986214\",\n\"NM_001244008.2(KIF1A):c.746T>C (p.Leu249Pro)\",\n\"KIF1A\",\n\"02\",\n\"00000000000240783791\",\n\"single nucleotide variant\",\n\"L249P\",\n\"Pathogenic/Likely pathogenic\"]],\n        columns: [[\"number\", \"index\"], [\"string\", \"ClinVar ID\"], [\"string\", \"title\"], [\"string\", \"gene_sort\"], [\"string\", \"chr_sort\"], [\"string\", \"location_sort\"], [\"string\", \"obj_type\"], [\"string\", \"protein_change\"], [\"string\", \"germline_classification\"]],\n        columnOptions: [{\"width\": \"1px\", \"className\": \"index_column\"}],\n        rowsPerPage: 25,\n        helpUrl: \"https://colab.research.google.com/notebooks/data_table.ipynb\",\n        suppressOutputScrolling: true,\n        minimumWidth: undefined,\n      });\n\n      function appendQuickchartButton(parentElement) {\n        let quickchartButtonContainerElement = document.createElement('div');\n        quickchartButtonContainerElement.innerHTML = `\n    <div id=\"df-ef1015b0-ea46-4070-b6c6-83c10ed26029\">\n      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-ef1015b0-ea46-4070-b6c6-83c10ed26029')\"\n                title=\"Suggest charts\"\n                style=\"display:none;\">\n\n<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n     width=\"24px\">\n    <g>\n        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n    </g>\n</svg>\n      </button>\n\n<style>\n  .colab-df-quickchart {\n      --bg-color: #E8F0FE;\n      --fill-color: #1967D2;\n      --hover-bg-color: #E2EBFA;\n      --hover-fill-color: #174EA6;\n      --disabled-fill-color: #AAA;\n      --disabled-bg-color: #DDD;\n  }\n\n  [theme=dark] .colab-df-quickchart {\n      --bg-color: #3B4455;\n      --fill-color: #D2E3FC;\n      --hover-bg-color: #434B5C;\n      --hover-fill-color: #FFFFFF;\n      --disabled-bg-color: #3B4455;\n      --disabled-fill-color: #666;\n  }\n\n  .colab-df-quickchart {\n    background-color: var(--bg-color);\n    border: none;\n    border-radius: 50%;\n    cursor: pointer;\n    display: none;\n    fill: var(--fill-color);\n    height: 32px;\n    padding: 0;\n    width: 32px;\n  }\n\n  .colab-df-quickchart:hover {\n    background-color: var(--hover-bg-color);\n    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n    fill: var(--button-hover-fill-color);\n  }\n\n  .colab-df-quickchart-complete:disabled,\n  .colab-df-quickchart-complete:disabled:hover {\n    background-color: var(--disabled-bg-color);\n    fill: var(--disabled-fill-color);\n    box-shadow: none;\n  }\n\n  .colab-df-spinner {\n    border: 2px solid var(--fill-color);\n    border-color: transparent;\n    border-bottom-color: var(--fill-color);\n    animation:\n      spin 1s steps(1) infinite;\n  }\n\n  @keyframes spin {\n    0% {\n      border-color: transparent;\n      border-bottom-color: var(--fill-color);\n      border-left-color: var(--fill-color);\n    }\n    20% {\n      border-color: transparent;\n      border-left-color: var(--fill-color);\n      border-top-color: var(--fill-color);\n    }\n    30% {\n      border-color: transparent;\n      border-left-color: var(--fill-color);\n      border-top-color: var(--fill-color);\n      border-right-color: var(--fill-color);\n    }\n    40% {\n      border-color: transparent;\n      border-right-color: var(--fill-color);\n      border-top-color: var(--fill-color);\n    }\n    60% {\n      border-color: transparent;\n      border-right-color: var(--fill-color);\n    }\n    80% {\n      border-color: transparent;\n      border-right-color: var(--fill-color);\n      border-bottom-color: var(--fill-color);\n    }\n    90% {\n      border-color: transparent;\n      border-bottom-color: var(--fill-color);\n    }\n  }\n</style>\n\n      <script>\n        async function quickchart(key) {\n          const quickchartButtonEl =\n            document.querySelector('#' + key + ' button');\n          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n          quickchartButtonEl.classList.add('colab-df-spinner');\n          try {\n            const charts = await google.colab.kernel.invokeFunction(\n                'suggestCharts', [key], {});\n          } catch (error) {\n            console.error('Error during call to suggestCharts:', error);\n          }\n          quickchartButtonEl.classList.remove('colab-df-spinner');\n          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n        }\n        (() => {\n          let quickchartButtonEl =\n            document.querySelector('#df-ef1015b0-ea46-4070-b6c6-83c10ed26029 button');\n          quickchartButtonEl.style.display =\n            google.colab.kernel.accessAllowed ? 'block' : 'none';\n        })();\n      </script>\n    </div>`;\n        parentElement.appendChild(quickchartButtonContainerElement);\n      }\n\n      appendQuickchartButton(table);\n    ",
            "text/html": [
              "\n",
              "  <div id=\"df-4ee40c7e-3592-43fa-aa7b-d2f28db27838\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ClinVar ID</th>\n",
              "      <th>title</th>\n",
              "      <th>gene_sort</th>\n",
              "      <th>chr_sort</th>\n",
              "      <th>location_sort</th>\n",
              "      <th>obj_type</th>\n",
              "      <th>protein_change</th>\n",
              "      <th>germline_classification</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>986214</td>\n",
              "      <td>NM_001244008.2(KIF1A):c.746T&gt;C (p.Leu249Pro)</td>\n",
              "      <td>KIF1A</td>\n",
              "      <td>02</td>\n",
              "      <td>00000000000240783791</td>\n",
              "      <td>single nucleotide variant</td>\n",
              "      <td>L249P</td>\n",
              "      <td>Pathogenic/Likely pathogenic</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-4ee40c7e-3592-43fa-aa7b-d2f28db27838')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-4ee40c7e-3592-43fa-aa7b-d2f28db27838 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-4ee40c7e-3592-43fa-aa7b-d2f28db27838');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "  <div id=\"id_b9bed6f0-3904-4450-a90b-bafef0a8a932\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('df')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_b9bed6f0-3904-4450-a90b-bafef0a8a932 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('df');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "  ClinVar ID                                         title gene_sort chr_sort  \\\n",
              "0     986214  NM_001244008.2(KIF1A):c.746T>C (p.Leu249Pro)     KIF1A       02   \n",
              "\n",
              "          location_sort                   obj_type protein_change  \\\n",
              "0  00000000000240783791  single nucleotide variant          L249P   \n",
              "\n",
              "        germline_classification  \n",
              "0  Pathogenic/Likely pathogenic  "
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# Search ClinVar for ClinVar IDs\n",
        "# This section can be used to find ClinVar IDs for mutations of interest. It can be run as many times as needed before starting analysis by clicking the \"Play\" button to the left. Before running for the first time, also execute the section above.\n",
        "email = 'liangus@broadinstitute.org'\n",
        "# An email address is required for Entrez searches for ClinVar records. We do not store your email.\n",
        "search_term = \"KIF1A L249P\"\n",
        "max_record_number = 5\n",
        "if search_term:\n",
        "  Entrez.email = email\n",
        "  handle = Entrez.esearch(db='clinvar', retmax = max(1, max_record_number), term=search_term)\n",
        "  record = Entrez.read(handle)\n",
        "  cvIds_search = record['IdList']\n",
        "  post = Entrez.read(Entrez.epost('clinvar', id=','.join(cvIds_search)))\n",
        "  webenv = post['WebEnv']\n",
        "  query_key = post['QueryKey']\n",
        "  records = Entrez.read(Entrez.esummary(db='clinvar', query_key=query_key, WebEnv=webenv), validate = False)\n",
        "  df = pd.DataFrame.from_dict(records['DocumentSummarySet']['DocumentSummary'])\n",
        "  df = df[['title','gene_sort','chr_sort','location_sort','obj_type','protein_change','germline_classification']]\n",
        "  df.insert(0, 'ClinVar ID', cvIds_search)\n",
        "  df['germline_classification'] = df['germline_classification'].apply(lambda x: x['description'])\n",
        "  display(df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "tkO5jMy1Tfg5",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "tkO5jMy1Tfg5",
        "outputId": "a288cfe0-b34d-41fd-82fd-0a3a29eb641a"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Run 1 of 2\n",
            "ClinVar ID: 30169\n",
            "Variant name: NM_001244008.2(KIF1A):c.296C>T (p.Thr99Met)\n",
            "WT:\n",
            "GGATACAACGTGTGCATCTTCGCCTATGGGCAGACGGGTGCCGGCAAGTCCTACACCATGATGGGCAAGC\n",
            " G  Y  N  V  C  I  F  A  Y  G  Q  T  G  A  G  K  S  Y  T  M  M  G  K \n",
            "CCTATGTTGCACACGTAGAAGCGGATACCCGTCTGCCCACGGCCGTTCAGGATGTGGTACTACCCGTTCG\n",
            "Variant:\n",
            "GGATACAACGTGTGCATCTTCGCCTATGGGCAGATGGGTGCCGGCAAGTCCTACACCATGATGGGCAAGC\n",
            " G  Y  N  V  C  I  F  A  Y  G  Q  M  G  A  G  K  S  Y  T  M  M  G  K \n",
            "CCTATGTTGCACACGTAGAAGCGGATACCCGTCTACCCACGGCCGTTCAGGATGTGGTACTACCCGTTCG\n",
            "The wild-type codon ACG, coding for T, is mutated to ATG, coding for M\n",
            "ATG could be corrected to ACG\n",
            "Sending GCTTGCCCATCATGGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCG (top strand) to BE-Hive with editor ABE8e\n",
            "Sending CTTGCCCATCATGGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGA (top strand) to BE-Hive with editor ABE8e\n",
            "Sending TTGCCCATCATGGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAA (top strand) to BE-Hive with editor ABE8e\n",
            "Sending TGCCCATCATGGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAG (top strand) to BE-Hive with editor ABE8e\n",
            "Sending GCCCATCATGGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAGA (top strand) to BE-Hive with editor ABE8e\n",
            "Sending CCCATCATGGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAGAT (top strand) to BE-Hive with editor ABE8e\n",
            "Sending CCATCATGGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAGATG (top strand) to BE-Hive with editor ABE8e\n",
            "Sending CATCATGGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAGATGC (top strand) to BE-Hive with editor ABE8e\n",
            "Sending ATCATGGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAGATGCA (top strand) to BE-Hive with editor ABE8e\n",
            "Sending TCATGGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAGATGCAC (top strand) to BE-Hive with editor ABE8e\n",
            "Sending CATGGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAGATGCACA (top strand) to BE-Hive with editor ABE8e\n",
            "Sending ATGGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAGATGCACAC (top strand) to BE-Hive with editor ABE8e\n",
            "Sending TGGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAGATGCACACG (top strand) to BE-Hive with editor ABE8e\n",
            "Sending GGTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAGATGCACACGT (top strand) to BE-Hive with editor ABE8e\n",
            "Sending GTGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAGATGCACACGTT (top strand) to BE-Hive with editor ABE8e\n",
            "Sending TGTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAGATGCACACGTTG (top strand) to BE-Hive with editor ABE8e\n",
            "Sending GTAGGACTTGCCGGCACCCATCTGCCCATAGGCGAAGATGCACACGTTGT (top strand) to BE-Hive with editor ABE8e\n",
            "Done\n",
            "Run 2 of 2\n",
            "ClinVar ID: 986214\n",
            "Variant name: NM_001244008.2(KIF1A):c.746T>C (p.Leu249Pro)\n",
            "WT:\n",
            "GTGAGCAAAATCAGCCTGGTGGACCTGGCTGGGAGCGAGCGGGCTGACTCCACGGGAGCCA\n",
            " V  S  K  I  S  L  V  D  L  A  G  S  E  R  A  D  S  T  G  A \n",
            "CACTCGTTTTAGTCGGACCACCTGGACCGACCCTCGCTCGCCCGACTGAGGTGCCCTCGGT\n",
            "Variant:\n",
            "GTGAGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCTGACTCCACGGGAGCCA\n",
            " V  S  K  I  S  L  V  D  P  A  G  S  E  R  A  D  S  T  G  A \n",
            "CACTCGTTTTAGTCGGACCACCTGGGCCGACCCTCGCTCGCCCGACTGAGGTGCCCTCGGT\n",
            "The wild-type codon CTG, coding for L, is mutated to CCG, coding for P\n",
            "CCG could be corrected to TTG\n",
            "Sending CTGTTTCAGGTGAGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCG (bottom strand) to BE-Hive with editor BE4\n",
            "Sending TGTTTCAGGTGAGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGG (bottom strand) to BE-Hive with editor BE4\n",
            "Sending GTTTCAGGTGAGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGG (bottom strand) to BE-Hive with editor BE4\n",
            "Sending TTTCAGGTGAGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGC (bottom strand) to BE-Hive with editor BE4\n",
            "Sending TTCAGGTGAGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCT (bottom strand) to BE-Hive with editor BE4\n",
            "Sending TCAGGTGAGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCTG (bottom strand) to BE-Hive with editor BE4\n",
            "Sending CAGGTGAGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCTGA (bottom strand) to BE-Hive with editor BE4\n",
            "Sending AGGTGAGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCTGAC (bottom strand) to BE-Hive with editor BE4\n",
            "Sending GGTGAGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCTGACT (bottom strand) to BE-Hive with editor BE4\n",
            "Sending GTGAGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCTGACTC (bottom strand) to BE-Hive with editor BE4\n",
            "Sending TGAGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCTGACTCC (bottom strand) to BE-Hive with editor BE4\n",
            "Sending GAGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCTGACTCCA (bottom strand) to BE-Hive with editor BE4\n",
            "Sending AGCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCTGACTCCAC (bottom strand) to BE-Hive with editor BE4\n",
            "Sending GCAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCTGACTCCACG (bottom strand) to BE-Hive with editor BE4\n",
            "Sending CAAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCTGACTCCACGG (bottom strand) to BE-Hive with editor BE4\n",
            "Sending AAAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCTGACTCCACGGG (bottom strand) to BE-Hive with editor BE4\n",
            "Sending AAATCAGCCTGGTGGACCCGGCTGGGAGCGAGCGGGCTGACTCCACGGGA (bottom strand) to BE-Hive with editor BE4\n",
            "Done\n"
          ]
        },
        {
          "data": {
            "application/javascript": "\n    async function download(id, filename, size) {\n      if (!google.colab.kernel.accessAllowed) {\n        return;\n      }\n      const div = document.createElement('div');\n      const label = document.createElement('label');\n      label.textContent = `Downloading \"${filename}\": `;\n      div.appendChild(label);\n      const progress = document.createElement('progress');\n      progress.max = size;\n      div.appendChild(progress);\n      document.body.appendChild(div);\n\n      const buffers = [];\n      let downloaded = 0;\n\n      const channel = await google.colab.kernel.comms.open(id);\n      // Send a message to notify the kernel that we're ready.\n      channel.send({})\n\n      for await (const message of channel.messages) {\n        // Send a message to notify the kernel that we're ready.\n        channel.send({})\n        if (message.buffers) {\n          for (const buffer of message.buffers) {\n            buffers.push(buffer);\n            downloaded += buffer.byteLength;\n            progress.value = downloaded;\n          }\n        }\n      }\n      const blob = new Blob(buffers, {type: 'application/binary'});\n      const a = document.createElement('a');\n      a.href = window.URL.createObjectURL(blob);\n      a.download = filename;\n      div.appendChild(a);\n      a.click();\n      div.remove();\n    }\n  ",
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "data": {
            "application/javascript": "download(\"download_5458fc1d-c402-4b69-a68c-6315d392cf98\", \"ColabBE 2025-06-12 20H 56M 51S.zip\", 151157)",
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# Perform analysis and download results\n",
        "# When all parameters have been specified, start analysis by clicking the \"Play\" button to the left. A CPU runtime is sufficient.\n",
        "email = 'liangus@broadinstitute.org' \n",
        "# An email address is required for Entrez searches for ClinVar records. We do not store your email.\n",
        "mode = 'Enter ClinVar IDs directly' \n",
        "# Inputs for modes other than the one selected will be ignored. Excel and CSV files are supported for upload. If applicable, you will be prompted to upload after clicking the \"Play\" button to the left.\n",
        "spreadsheet_has_headers = True \n",
        "# If you are uploading a spreadsheet with column headers, select `spreadsheet_has_headers` and enter the name of the column with ClinVar IDs here.\n",
        "spreadsheet_column = 'VariationID' \n",
        "# If you are uploading a spreadsheet with column headers, specify the name of the column containing ClinVar IDs here. If your spreadsheet does not contain headers, the first column of the spreadsheet will automatically be used to retrieve the list of ClinVar IDs.\n",
        "gene = 'ATM' \n",
        "# To search for corrective base editing strategies to all pathogenic/likely pathogenic mutations in a gene documented in ClinVar (may take a long time), use the \"Enter a gene name\" mode.\n",
        "clinVar_ids = '30169, 986214' \n",
        "# ClinVar IDs found using the search form above can be entered here. ClinVar IDs can be separated by any mix of spaces, commas, and/or semicolons. This list will be used if the \"Enter ClinVar IDs directly\" mode is selected.\n",
        "average_editing_efficiency = 0.5 # @param {type:\"slider\", min:0.01, max:0.99, step:0.01}\n",
        "# Specify the average editing efficiency of your base editor in `average_editing_efficiency`. This only affects the predicted editing efficiency at each target and will not alter the list of suggested editing strategies.\n",
        "threshold = 0.6 # @param {type:\"slider\", min:0.01, max:0.99, step:0.01}\n",
        "# Use `threshold` to specify the cutoff threshold for predicted edit purity required for inclusion in the abbreviated summary of high-potential editor candidates. For coding mutations, purity is measured by exact match to reference translated protein sequence. For non-coding mutations, purity is measured by exact DNA sequence match to reference. This does not alter the full editor analysis that is also provided.\n",
        "verbose_output = True # @param {type:\"boolean\"}\n",
        "# Selecting `verbose_output` will provide a running summary of edit strategies being tested. If deselected, only the current ClinVar ID and variant name being tested will be displayed.\n",
        "\n",
        "if mode == \"Enter ClinVar IDs directly\":\n",
        "  cvIds = re.split(r'[ ,;\\t]+', clinVar_ids)\n",
        "elif mode == \"Enter a gene name\":\n",
        "  Entrez.email = email\n",
        "  handle = Entrez.esearch(db='clinvar', retmax = 5000, term=f'(((\"{gene}\"[Gene Name]) AND \"single nucleotide variant\"[Type of variation]) AND (\"clinsig pathogenic\"[Filter] OR \"clinsig likely path\"[Filter]))')\n",
        "  record = Entrez.read(handle)\n",
        "  cvIds = record['IdList']\n",
        "elif mode == \"Upload a spreadsheet containing ClinVar IDs\":\n",
        "  fname = ''\n",
        "  ext = fname.split('.')[-1]\n",
        "  if ext == 'xlsx' or ext == 'xls':\n",
        "    if not spreadsheet_has_headers:\n",
        "      df = pd.read_excel(fname, header = None)\n",
        "      cvIds = df.iloc[:, 0].to_list()\n",
        "    else:\n",
        "      df = pd.read_excel(fname)\n",
        "      cvIds = df[spreadsheet_column].to_list()\n",
        "  elif ext == 'csv':\n",
        "    if not spreadsheet_has_headers:\n",
        "      df = pd.read_csv(fname, header = None)\n",
        "      cvIds = df.iloc[:, 0].to_list()\n",
        "    else:\n",
        "      df = pd.read_csv(fname)\n",
        "      cvIds = df[spreadsheet_column].to_list()\n",
        "  else: raise Exception('Unsupported file type')\n",
        "  cvIds = [str(cvId) for cvId in cvIds]\n",
        "elif mode == \"Use all ClinVar search results from above\":\n",
        "  try: cvIds = cvIds_search\n",
        "  except: print('No ClinVar search was run in the current session. Run a ClinVar search and try again.')\n",
        "\n",
        "\n",
        "homedir = '/content/gdrive/MyDrive'\n",
        "timestamp = datetime.now().strftime('%Y-%m-%d %HH %MM %SS')\n",
        "os.mkdir(os.path.join(homedir,f'ColabBE {timestamp}'))\n",
        "run = 1\n",
        "candidates = []\n",
        "hits = {}\n",
        "for cvId in cvIds:\n",
        "    print(f'Run {run} of {len(cvIds)}')\n",
        "    [sequence_distributions, stats_dict, vname] = analyze(cvId, verbose_output, average_editing_efficiency)\n",
        "    if stats_dict:\n",
        "        with pd.ExcelWriter(os.path.join(homedir,f'ColabBE {timestamp}',f'{cvId} ({vname}) predicted sequence distributions.xlsx')) as seq_dist:\n",
        "            for entry in sequence_distributions:\n",
        "                (editor, spacer, pam) = entry\n",
        "                df_seq = pd.DataFrame.from_dict(sequence_distributions[entry], orient='index', columns=['Frequency'])\n",
        "                df_seq.to_excel(seq_dist, sheet_name=f'{editor} {spacer}|{pam}', index_label='Sequence')\n",
        "        for entry in stats_dict:\n",
        "            if entry['PAM'] in pam_dict:\n",
        "              entry['HT-PAMDA score'] = pam_dict[entry['PAM']]['max_score']\n",
        "              entry['Recommended Cas9 variant'] = pam_dict[entry['PAM']]['best_PAM']\n",
        "            else:\n",
        "              entry['HT-PAMDA score'] = '< -2.5'\n",
        "              entry['Recommended Cas9 variant'] = 'Non-SpCas9'\n",
        "            if entry['Perfect correction'] >= threshold:\n",
        "                candidates.append(entry)\n",
        "                hits[entry['ClinVar ID']] = entry['Variant name']\n",
        "        df_stats = pd.DataFrame.from_dict(stats_dict)\n",
        "        df_stats.to_excel(os.path.join(homedir,f'ColabBE {timestamp}',f'{cvId} ({vname}) predicted base editing statistics.xlsx'))\n",
        "    run += 1\n",
        "df_summary = pd.DataFrame.from_dict(candidates)\n",
        "df_summary.to_excel(os.path.join(homedir,f'ColabBE {timestamp}',f'Summary of editor candidates above threshold.xlsx'))\n",
        "df_hits = pd.DataFrame.from_dict(hits, orient='index', columns=['Variant name'])\n",
        "df_hits.to_excel(os.path.join(homedir,f'ColabBE {timestamp}',f'Variants with editor candidates above threshold.xlsx'), index_label='ClinVar ID')\n",
        "subprocess.run(['zip', '-r', os.path.join(homedir, f'ColabBE {timestamp}.zip'), os.path.join(homedir, f'ColabBE {timestamp}')])"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "mlfold",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.12.7"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
